### PR TITLE
Add options for multilanguage strings in the installer

### PR DESCRIPTION
**Pull request by bovirus**

### CODE REVIEW

File: PowerEditor/installer/nsisInclude/langs4Installer.nsh [hunk 1]
```diff
@@ -153,3 +153,237 @@ LangString langFileName ${LANG_GEORGIAN} "georgian.xml"
...
+LangString LocalizationWelshItem ${LANG_English} "Welsh"
+LangString LocalizationKurdishItem ${LANG_English} Kurdish"
+LangString LocalizationPiglatinItem ${LANG_English} "Piglatin"
+LangString LocalizationZuluItem ${LANG_English} "Zulu"
+LangString CreateShortcutonDesktop ${LANG_English} "Create Shortcut on Desktop"
+LangString DontUseAppdata ${LANG_English} "Don't use %APPDATA%$\nEnable this option to make Notepad++ load/write the configuration files from/to its install directory. Check it if you use Notepad++ in a USB device."
+LangString NotepadNotSuportThisOS ${LANG_English} "This version of Notepad++ does not support your OS.$\nPlease download zipped package of version 5.9 and use ANSI version. You can find v5.9 here:$\nhttp://notepad-plus-plus.org/release/5.9"
+LangString CannotContinueInstallation ${LANG_English}"Cannot continue the installation: Notepad++ is running.\$\n$\n\Please close Notepad++, then click ''Retry''."
+
+; Language Strings (Italian)
```
Comment (donho) [inline PowerEditor/installer/nsisInclude/langs4Installer.nsh]: It means all the translations should go here - it will make this file so huge.

**Comment by donho** [comment]
Make Notepad++ (NSIS) installer completely translatable is not a trivial job. So I prefer to not start (and then maintain) this part.

**Comment by bovirus** [comment]
Sorry but I didn't understand what's your problem to have a program localized.
The job is already done and you don't have to spend 1 minutes to do that (it's already did).

I spent a lot of my spare time to do this job for you (because I like Notepad++ and don't like to see a mix of Italian and English strings during installation like now).

Then what's your problem?
How many resources/time are requested to you to do that? None.
And finally you will have a really multi-language installer.
You just add the basic language files for NSIS.
But you don't translate the description for component/sections.
I did it for you but (nothing requested to you as time resources) and you say "closed"
What's about?

**Comment by donho** [comment]
@bovirus 
From the aspect of Notepad++ project, a generic solution is needed, but what you have done is **only** Italian localization. If we want to have ALL the supported translations, all the translation of all languages should be added into `langs4Installer.nsh` - that will make this file so huge and difficult to maintain.

What's my problem? My problem is I have to decide which direction is good for this project, and take only the "right" PR.

I understand that you have taken your time for your PR. Please think of my time for offering Notepad++ as open source project under GPL since 2003! It allows anyone, include you to fork this project, integrate your code and distribute it as you want (with the modified source of course).

As Notepad++ project lead and main maintainer, I think my decision is good enough to bring this project to where it is. OTOH you can always fork this project to merge your PR and distribute it, fair enough?

**Comment by bovirus** [comment]
English is base. 
Italian is new installer language and is first brick of the wall (a lot of brick = languages).
If this option will be available I believe that the other translator comes in.

Sorry I didn't understand why didn't accept the PR  that offer new function at your product without required additional time to implement it.

I would't make a  fork.
I would that my favorite editor and its official fork has a really multi-language installer.
I develop it and you refuse it but I didn't understand really why and what's are the negative point for that.

I believe that you can implement it, declare it in changelog or you like and I believe that in a short time as you found a lot of translator for your program GUI you will find translator for the installer strings.

Thanks. 
The other translator cannot do it before in the original script there no option to localzie teh strings.-
Now the start track to create new localziation are in place and any of other translator can add

**Comment by bovirus** [comment]
@donho 

I ask to reconsider pelase the reject of this prosal.
For me it's an interesting plus for all translator.

I didn't udnesratnd because you reject it. No negative points for you.
For you don't have any neagtive point and you don't have to spent resources to add that (already done!!!).

@rddim 
@Tmp341 
@Patricollu
@SinghRajenM 
@xylographe 

Please check if could be interested for you.

I believe that could be inteersted for you and other guys that use local language to have a fully translated installer. 
I know for ex. that in France the user have exptected to have the program much translated as possible (also in the installer).

**Comment by Tmp341** [comment]
Well, from the perspective of a Translator, i don't actually care as long as it is not a 2k+ lines of  translation. I can do that too, but too much time will be needed.

From the perspective of a user, who don't know English very well, i think it will be a good option. But, installer isn't too much complicated, so i think users won't have too much problem anyway, but like i said, it would be a good option.

**Comment by donho** [comment]
I didn't say there won't be any contributor for translating it.
I was saying that the maintaining job is fastidious:
> we have to add the translations of all languages into langs4Installer.nsh - that will make this file so huge and difficult to maintain.

The fact every translator will modify the same file, I have to resolve the conflict each time, or I have to integrate the translation manually each time there's a modification or a new entry.

Currently NSIS multilingual system covers 90% of localization in Notepad++ installer, for me it's good enough for an installer, and I'm not ready to start my catastrophic life for making just 10% part of installer translatable.

Not mentioning `langs4Installer.nsh` will become huge and ugly - there's just no way that I integrate this PR. Hope I'm clear.

**Comment by SinghRajenM** [comment]
@donho @bovirus 
> The fact every translator will modify the same file, I have to resolve the conflict each time, or I have to integrate the translation manually each time there's a modification or a new entry.

I think this part can be improved. We can have dedicated string file for each lang. So there will be no conflict as such. Moreover, this will be one time activity as installer changes rarely for strings (localization).

For example:
image

**Comment by bovirus** [comment]
@Tmp341 

Please check.
The line to translate are exactly 122. For each language. Not 2k+.

@donho 

The current problem is the problem that you explained.

The big problem is with current scheme you have most of the string hardcoded as English in the installer and the translator that want to translate them have to change a lot of file just for specfic language.

With my proposal is more easy and flexible.

For me in the perspective of flexible maintening is quite simple.

Please take care that most of the strings are already replaced by me.

You as master maintener just if he want to add new strings add not as harcoded text but as link to the master Eeglish string in

langs4Installer.nsh

and point the relative string name with $ ad I did for current strings.

Then you maintain the source code and avoid problem in PR on source code.

The translators didn't touch the source code and just revise the specific translation section just touch

langs4Installer.nsh

and you have to check only this file for PR by translator.

I didn't uednrstand what's the problem about the lenght if 

langs4Installer.nsh

it will be not GB as lenght. Could be some KB.

I already applied these scheme to other project with NSISI and I get a lot time saved with this way.

**Comment by donho** [comment]
What @SinghRajenM has suggested could be an accepted solution - a nsh file per language.
My preference is a little different: 
Create `localization` folder under `PowerEditor/installer/nsisInclude/` then we could use the same naming schema of `PowerEditor/installer/nativeLang/` to have:
`PowerEditor/installer/nsisInclude/localization/arabic.nsh`
`PowerEditor/installer/nsisInclude/localization/chinese.nsh`
`PowerEditor/installer/nsisInclude/localization/english.nsh`
`PowerEditor/installer/nsisInclude/localization/french.nsh`
`PowerEditor/installer/nsisInclude/localization/italian.nsh`
...

then in `langs4Installer.nsh` just include them.

@bovirus Could you try this solution?

**Comment by Tmp341** [comment]
> 
> @Tmp341
> 
> Please check.
> The line to translate are exactly 122. For each language. Not 2k+.

I was just stating that ``If a language file contains too much lines, it would be hard for me (and probably for other translators too)``. I was talking generally, i know it does not contain 2k+ lines.

**Comment by yasmise** [comment]
@bovirus If translation file is separated into each languages, as proposed by @SinghRajenM, I think I could try to translate. But I don't want to translate `langs4Installer.nsh` directly, as it may contain other script lines which are nothing related for translation. In other words, it's a little bit hard for me to tell which lines are to be translated.

By the way, I have two questions. Could you take a look?
1. Are there any way to check my translated file is correct, other than building whole Notepad++ project?  
I'm afraid that my wrong translation file could break installer. (Please note that I'm just a volunteer translator and not a software enginner.)

1. What happens when new English texts is added, but texts of other languages are not added?  
On current translation framework for Notepad++ itself, if newly added texts are not translated yet, they are automatically fallback to English. Is it same for this PR?

**Comment by bovirus** [comment]
@SinghRajenM 

Thanks for the proposal. It's ok for me. See later the request for the notepad++ developer.

@yasmise 

1. Which is the problem of your translation?
2. If a string is not availble for a specific language should be sued the master English string.

@donho 
For me the system proposed by @SinghRajenM is OK. 

My main scope is to have a flexible system that enable after few changes in main scripts in source code to have as much possible the strings used in installer translated (if translation is available).

Then the system proposed bySinghRajenM get my scope.

What can I check to do this?

**Comment by yasmise** [comment]
@bovirus Thanks for your reply.
>     1. Which is the problem that you found on your translation?
Sorry for my bad explanation. What I mean is: I'm appreciate if I could easily check my translation file before making a pull request in the future.

If I understand correctly, your new translation files are config files for the installer. I'm thinking of these could happen (just a thought, so maybe wrong):
* If I made a bad translation file which contains syntax errors, and unfortunately merged, installer could stop working.
* If I made a bad translation file which contains too long string to fit in installer's GUI, users can't read the whole text.

I want to check by myself to avoid issues mentioned above in the future. But I don't know how to do, so, I'm appreciate if you could tell me.

**Comment by donho** [comment]
@yasmise has revealed a very good point:
> What I mean is: I'm appreciate if I could easily check my translation file before making a pull request in the future.

So I doubt the quality of commit because it's not verified even by the translators. 
In case of the problem (installer build fails), it'll need integrater's effort to find the problem, fix it and do the release.

**Comment by bovirus** [comment]
@bovirus 

Using the propsoal of @doho how could the structure syntax of languae.nsh file

Ex. I want to change the file **_PowerEditor\installer\nsisIclude\Autocompletion.nsh_** text strings 

SectionGroup **_"Auto-completion Files"_** autoCompletionComponent

with

SectionGroup **_${AutoCompletionFilesGroup}_** autoCompletionComponent

what should be the syntax in **_PowerEditor/installer/nsisInclude/localization/English.nsh_**

LangString AutoCompletionFilesSection ${LANG_English} "Auto-completion Files"

and for talian file  **PowerEditor/installer/nsisInclude/localization/Italian.nsh**

LangString AutoCompletionFilesSection ${LANG_Italian} "File au completamento"

**Comment by xylographe** [comment]
Perhaps, use `!addincludedir` ?
```
!addincludedir ${NSISDIR}\nsisInclude\localization

!include english.nsh
!include french.nsh
!include chinese.nsh
!include chineseSimplified.nsh
!include korean.nsh
!include japanese.nsh
!include german.nsh
!include spanish.nsh
!include italian.nsh
!include portuguese.nsh
!include brazilian_portuguese.nsh
!include dutch.nsh
!include russian.nsh
!include polish.nsh
!include catalan.nsh
!include czech.nsh
!include hungarian.nsh
!include romanian.nsh
!include turkish.nsh
!include farsi.nsh
!include ukrainian.nsh
!include hebrew.nsh
!include hindi.nsh
!include nynorsk.nsh
!include norwegian.nsh
!include thai.nsh
!include arabic.nsh
!include finnish.nsh
!include lithuanian.nsh
!include greek.nsh
!include swedish.nsh
!include galician.nsh
!include slovenian.nsh
!include slovak.nsh
!include danish.nsh
!include bulgarian.nsh
!include indonesian.nsh
!include albanian.nsh
!include croatian.nsh
!include basque.nsh
!include belarusian.nsh
!include serbian.nsh
!include malay.nsh
!include luxembourgish.nsh
!include afrikaans.nsh
!include uzbek.nsh
!include macedonian.nsh
!include Latvian.nsh
!include bosnian.nsh
!include mongolian.nsh
!include estonian.nsh
!include corsican.nsh
!include breton.nsh
!include georgian.nsh
!include vietnamese.nsh
!include welsh.nsh
!include kurdish.nsh
```

**Comment by bovirus** [comment]
@xylographe

thanks for the idea. I will check it.

**Pull request closed without merging**